### DISCUSSION TITLE
Declare external reactions in C++ code, rather than in XML configuration.

### DISCUSSION THREAD

**Discussion by scotteh432** [JSBSim-Team/jsbsim#1204] [Q&A]
Hi all,

I'm embedding JSBSim inside another program for a project I am working on involving aircraft towing systems. To get this to work, it would be very nice to be able to declare an External Reaction via the C++ code rather than in the XML configuration files. I've read through the documentation on the FGExternalReactions, which indicates that this should be possible, but alas, I cannot figure out how. 

I've tried accessing the FGExternalReactions through the GetExternalReactions from the FGFDMExec class. But I can't seem to do anything with it once I have it without causing my code to not compile. I believe I should be able to call the Load() method to initialize an external reaction from an XML element. But I can't seem to call this method without causing my code to fail to compile. 

```c++
// Setup directories for JSBSim execution
JSBSim::FGFDMExec> f;
f.SetRootDir(MyRootDir);
f.SetAircraftPath(MyAircraftDir);
f.SetEnginePath(MyEngineDir);
f.SetSystemsPath(MySystemsDir);
f.SetOutputPath(MyOurputDir);

// Ok, here! How would I declare an External Reaction?
shared_ptr<JSBSim::FGExternalReactions> ptr = f.GetExternalReactions();   // Is this a step in right direction? 
ptr->Load(someXmlElement);  // This does not work, my compiler complains I'm doing something wrong.

// Load script and launch init
f.LoadScript(MyScriptDir);
f.RunIC();

// Execution loop is down here; omitted for this snippet of code posted to the JSBSim forum
```

Any help in solving this issue would be much appreciated. I'm willing to accept the answer of "just use the XML configuration files" if there's no good way to do this, I just want to explore this avenue first, as I believe it will make the workflow of my project cleaner.  Thank you very much for your time.

**Comment by bcoconni** [comment]
```c++
// Ok, here! How would I declare an External Reaction?
shared_ptr<JSBSim::FGExternalReactions> ptr = f.GetExternalReactions();   // Is this a step in right direction? 
ptr->Load(someXmlElement);  // This does not work, my compiler complains I'm doing something wrong.
```

These lines look good to me however you should call them after your model has been loaded i.e. after `f.LoadScript(MyScriptDir)`.

If that does not help, it would be helpful to provide the error message that your compiler is returning.

**Comment by scotteh432** [comment]
The issue I was having was with my #include statements. I had to be ensure to have `FGExternalReactions.h `included. Thank you for the tip about calling Load only after calling `f.LoadScript(MyScriptDir)`, as that was also causing problems. I now have a different issue. As far as I'm aware, my dynamically added external reaction is working properly. My results appear intuitively correct, however I am receiving a warning message from JSBSim suggesting otherwise. For every external reaction created this way, I receive:

```
Failed to tie property moments/l-external-lbsft to indexed object methods
Failed to tie property moments/m-external-lbsft to indexed object methods
Failed to tie property moments/n-external-lbsft to indexed object methods
Failed to tie property forces/fbx-external-lbs to indexed object methods
Failed to tie property forces/fby-external-lbs to indexed object methods
Failed to tie property forces/fbz-external-lbs to indexed object methods
```

Here's what I am doing with my code so that I can access the external reaction location, vector, and magnitude, and modify it during the simulation in accordance with my simulation:

```
  // I have some variables in here that are declared above this block. For clarity
  // fdm is the JSBSim::FGFDMExec*
  // name is a string used to identify the external reaction. 
  // ele is the JSBSim::Element* used to create the external reaction
  std::shared_ptr<JSBSim::FGPropertyManager> m = fdm->GetPropertyManager();
  shared_ptr<JSBSim::FGExternalReactions> ptr = fdm->GetExternalReactions();
  ptr->Load(ele);

  // All these variables are automatically created when the external reaction is loaded. 
  // Therefore, I can pass create=false to GetNode. 
  // I am declaring variables to these nodes so I can more easily modify the values in the execution loop. 
  xLocationInches = m->GetNode("external_reactions/" + name + "/location-x-in", false);
  yLocationInches = m->GetNode("external_reactions/" + name + "/location-y-in", false);
  zLocationInches = m->GetNode("external_reactions/" + name + "/location-z-in", false);

  xDirectionVector = m->GetNode("external_reactions/" + name + "/x", false);
  yDirectionVector = m->GetNode("external_reactions/" + name + "/y", false);
  zDirectionVector = m->GetNode("external_reactions/" + name + "/z", false);

  // But for some reason magnitude is not created when the element is loaded?
  // And has to be created by me using create="true".  
  // Modifying this does successfully change the magnitude of the external reaction.
  // I don't understand this very well, and I suspect has something to do with my error. 
  magnitudeLbs = m->GetNode("external_reactions/" + name + "/magnitude", true);
```

My simulation appears to be doing what I want, but I'm wary of trusting the results when I'm being given warning messages like this. Any ideas on what the the warnings could be about, and how to fix them?
